flake.nix

```diff
@@ -253,11 +287,76 @@
           nil
           nixpkgs-fmt
           llvmPackages.lldb # This ensures the right lldb is in the environment for running rust-lldb
+          wasm-bindgen-cli
         ];
 
         shellHook = ''
+          echo which wasm-bindgen $(which wasm-bindgen)
           eval "$(starship init bash)"
         '';
       });
+
+      packages.wasm = craneLib.buildPackage rec {
+        pname = "noir_wasm";
+        version = "1.0.0";
+
+        inherit GIT_COMMIT;
+        inherit GIT_DIRTY;
+        doCheck = false;
+
+        COMMIT_SHORT = builtins.substring 0 7 GIT_COMMIT;
+        VERSION_APPENDIX = if GIT_DIRTY == "true" then "-dirty" else "";
+        PKG_PATH = "./pkg";
+
+        src = ./.;
+
+        nativeBuildInputs = with pkgs; [
+          which
+          git
+          jq
+          rustToolchain
+          wasm-bindgen-cli
+          binaryen
+          toml2json
+        ];
+
+        cargoExtraArgs = "--lib --package noir_wasm --target wasm32-unknown-unknown";
+
+        postBuild = ''
+          # Clear out the existing build artifacts as these aren't automatically removed by wasm-pack.
+          if [ -d ./pkg/ ]; then
+              rm -rf ./pkg/
+          fi
+          wasm-bindgen ./target/wasm32-unknown-unknown/release/noir_wasm.wasm --out-dir ./pkg/nodejs --typescript --target nodejs
+          wasm-bindgen ./target/wasm32-unknown-unknown/release/noir_wasm.wasm --out-dir ./pkg/web --typescript --target web
+          wasm-opt ./pkg/nodejs/noir_wasm_bg.wasm -o ./pkg/nodejs/noir_wasm_bg.wasm -O
+          wasm-opt ./pkg/web/noir_wasm_bg.wasm -o ./pkg/web/noir_wasm_bg.wasm -O
+        '';
+
+        installPhase = ''
+          # Extract version from Cargo.toml using toml2json
+          PACKAGE_VERSION=$(toml2json < Cargo.toml | jq -r .package.version)
+          if [ -z "$PACKAGE_VERSION" ]; then
+              echo "Could not extract version from Cargo.toml"
+              exit 1
+          fi
+          PACKAGE_VERSION+=$VERSION_APPENDIX
+
+          mkdir -p $out
+          cp README.md $out/
+          cp -r ./pkg/* $out/
+          jq -n --arg ver "$PACKAGE_VERSION" \
+              '{
+                "version": $ver, 
+                "repository": {"type": "git","url": "https://github.com/noir-lang/noir_wasm.git"},
+                "sideEffects": false, 
+                "files": ["nodejs","web","package.json"], 
+                "main": "./nodejs/noir_wasm.js", 
+                "types": "./web/noir_wasm.d.ts",
+                "module": "./web/noir_wasm.js"
+              }' > $out/package.json
```

Comment:
Would we rather have these as separate bash files?

Reply:
I tried doing this, and created an issue to track this: https://github.com/noir-lang/noir/issues/1544

I haven't been able to add external .sh files. I tried adding an shFilter to the srcFilter but it didn't work. I paired with Blaine, timeboxed it, and thought I might spending too much time on this. We can add that later on if we wish.

Reply:
Already sorted out with https://github.com/noir-lang/noir/pull/1657